### PR TITLE
add suport to aliyun-oss

### PR DESCRIPTION
**Pull request by luolibin**

### CODE REVIEW

File: oss/oss.go [hunk 1]
```diff
@@ -0,0 +1,779 @@
...
+	}
+	// read mtime out of metadata if available
+	d, ok := o.meta[metaMtime]
+	if !ok || d == nil {
+		// fs.Debugf(o, "No metadata")
+		return o.lastModified
+	}
+	modTime, err := swift.FloatStringToTime(*d)
+	if err != nil {
+		fs.Logf(o, "Failed to read mtime from object: %v", err)
```
Comment (xerrni) [inline oss/oss.go]: Failed to read modification time for object:

File: oss/oss.go [hunk 2]
```diff
@@ -0,0 +1,779 @@
...
+	if o.size >= maxSizeForCopy {
+		fs.Debugf(o, "SetModTime is unsupported for objects bigger than %v bytes", fs.SizeSuffix(maxSizeForCopy))
+		return nil
+	}
+	// Guess the content type
+	mimeType := fs.MimeType(o)
+	// Copy the object to itself to update the metadata
+	key := o.fs.root + o.remote
+	sourceKey := o.fs.bucket + "/" + key
+	bucket, err := o.fs.c.Bucket(o.fs.bucket)
```
Comment (xerrni) [inline oss/oss.go]: check if error occurred.

File: oss/oss.go [hunk 3]
```diff
@@ -0,0 +1,779 @@
...
+	_, erro := bucket.CopyObject(key, key, oss.ContentType(mimeType),
+		oss.CopySource(o.fs.bucket, url.QueryEscape(sourceKey)),
+		oss.MetadataDirective(oss.MetaReplace))
+	return erro
+}
+
+// Storable raturns a boolean indicating if this object is storable
```
Comment (xerrni) [inline oss/oss.go]: Storable returns a boolean indicating if the object is storable.

File: oss/oss.go [hunk 4]
```diff
@@ -0,0 +1,779 @@
...
+
+// Update the Object from in with modTime and size
+// 使用metaData更新对象
+func (o *Object) Update(in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) error {
+	err := o.fs.Mkdir("")
+	if err != nil {
+		return err
+	}
+	mimeType := fs.MimeType(src)
+	bucket, err := o.fs.c.Bucket(o.fs.bucket)
```
Comment (xerrni) [inline oss/oss.go]: check if error occurred

File: oss/oss.go [hunk 5]
```diff
@@ -0,0 +1,779 @@
...
+	// Read the metadata from the newly created object
+	o.meta = nil // wipe old metadata
+	erro = o.metaData(o.fs)
+	return erro
+}
+
+// Remove an object
+// 删除一个object
+func (o *Object) Remove() error {
+	bucket, _ := o.fs.c.Bucket(o.fs.bucket)
```
Comment (xerrni) [inline oss/oss.go]: check if error occurred

File: oss/oss.go [hunk 6]
```diff
@@ -0,0 +1,779 @@
...
+func (o *Object) MimeType() string {
+	err := o.metaData(o.fs)
+	if err != nil {
+		fs.Logf(o, "Failed to read metadata: %v", err)
+		return ""
+	}
+	return o.mimeType
+}
+
+// String returns a pointer to the string value passed in.
```
Comment (xerrni) [inline oss/oss.go]: StringToPointer returns...
Why you define function for this operation. 
It`s standard go operation.

File: oss/oss.go [hunk 7]
```diff
@@ -0,0 +1,779 @@
...
+// util 把string转换成string的指针
+func StringToPointer(v string) *string {
+	return &v
+}
+
+// Check the interfaces are satisfied
```
Comment (xerrni) [inline oss/oss.go]: Check if the interfaces are satisfied.

File: oss/oss.go [hunk 8]
```diff
@@ -0,0 +1,779 @@
...
+	bucketOK           bool         // true if we have created the bucket
+	locationConstraint string       // location constraint of new buckets
+	storageClass       string       // storage class
+	acl                string       // bucket acl
+}
+
+// Object describes a oss object
+type Object struct {
+	// Will definitely have everything but meta which may be nil
+	// List will read everything but meta & mimeType - to fill
```
Comment (xerrni) [inline oss/oss.go]: Correct comment. I don`t know what you mean. It`s not written in English.
Do the same for comment below.

**Comment by luolibin** [comment]
@xerrni I will modifiy the code ,thank you very much.

**Pull request closed without merging**